main/src/library/MutHashMap.flix

```diff
@@ -0,0 +1,817 @@
+    /// The capacity is rounded up to the minimum capacity.
+    ///
+    pub def emptyWithCapacity(rc: Region[r], capacity: Int32): MutHashMap[k, v, r] \ r =
+        let cap = Int32.max(capacity, minCapacity());
+        let buckets = Array.repeat(rc, cap, None);
+        new MutHashMap @ rc { rc = rc, buckets = buckets, head = None, tail = None, size = 0 }
+
+    ///
+    /// Returns a mutable hash map with a single key-value pair `k => v`.
+    ///
+    pub def singleton(rc: Region[r], k: k, v: v): MutHashMap[k, v, r] \ r with Eq[k], Hash[k] =
+        let m = empty(rc);
+        put(k, v, m);
+        m
+
+    ///
+    /// Returns the set of all keys in `m`.
+    ///
+    pub def keysOf(m: MutHashMap[k, v, r]): Set[k] \ r with Order[k] =
+        foldLeftWithKey((acc, k, _) -> Set.insert(k, acc), Set.empty(), m)
+
+    ///
+    /// Returns a list of all values in `m` in insertion order.
+    ///
+    pub def valuesOf(m: MutHashMap[k, v, r]): List[v] \ r =
+        foldRight((v, acc) -> v :: acc, Nil, m)
+
+    ///
+    /// Returns `Some(v)` if key `k` exists in `m`, otherwise `None`.
+    ///
+    pub def get(k: k, m: MutHashMap[k, v, r]): Option[v] \ r with Eq[k], Hash[k] =
+        let hash = Hash.hash(k);
+        let idx = hash2bucket(hash, Array.length(m->buckets));
+        match findInBucket(k, Array.get(idx, m->buckets)) {
+            case None => None
+            case Some(entry) => Some(Entry.getValue(entry))
+        }
+
+    ///
+    /// Returns the value associated with key `k` in `m`.
+    ///
+    /// If `k` does not exist, inserts `k` with value `default` and returns `default`.
+    ///
+    pub def getOrElsePut(k: k, default: v, m: MutHashMap[k, v, r]): v \ r with Eq[k], Hash[k] =
+        match get(k, m) {
+            case Some(v) => v
+            case None =>
+                put(k, default, m);
+                default
+        }
+
+    ///
+    /// Returns the value associated with key `k`, or `default` if not found.
+    ///
+    pub def getWithDefault(k: k, default: v, m: MutHashMap[k, v, r]): v \ r with Eq[k], Hash[k] =
+        Option.getWithDefault(default, get(k, m))
+
+    ///
+    /// Converts hash code to a valid bucket index.
+    /// Handles negative hashes correctly using bitwise AND.
+    ///
+    def hash2bucket(hash: Int32, numBuckets: Int32): Int32 =
+        Int32.bitwiseAnd(hash, numBuckets - 1)
+
+    ///
+    /// Searches for an entry with the given key in the bucket chain.
+    ///
+    def findInBucket(k: k, bucket: Option[Entry[k, v, r]]): Option[Entry[k, v, r]] \ r with Eq[k] =
+        match bucket {
+            case None => None
+            case Some(entry) =>
+                if (k == Entry.getKey(entry))
+                    Some(entry)
+                else
+                    findInBucket(k, Entry.getBucketNext(entry))
+        }
+
+    ///
+    /// Returns `true` if key `k` exists in `m`.
+    ///
+    pub def memberOf(k: k, m: MutHashMap[k, v, r]): Bool \ r with Eq[k], Hash[k] =
```

Comment:
don't we have `pub` before other methods generally?

Reply:
Ah yes. I tried to figure that out too. It seems somewhat inconsistent. In other modules the private functions sometimes come right after they are used. I know we have been using that scheme too, so I kept it here.